ROS 2 and microRTPS: improve documentation (#1014)
PX4/PX4-user_guide: en/ros/ros2_comm.md

# PX4-ROS2 bridge

> This section of the guide focusses on the bridge between PX4 and ROS 2, made available through the *microRTPS* bridge and resulting products to interface with ROS 2, *px4_ros_com* and *px4_msgs* ROS 2 packages.

Reviewer: This sentence doesn't make sense to me.

What I think it says is that the "ROS2- bridge" is the RTPS Bridge + a set of ROS2 pages named *px4_ros_com* and *px4_msgs*. 
Its a bit confusing where the separation is, because aren't those packages built as part of the  RTPS bridge?

Anyway, a little more clarification, possibly in both docs, to make it clear the boundary/diferrences/purposes of the two bridges might help.

Author: PX4-RTPS bridge aka microRTPS bridge = FastRTPS bridge with a client in PX4 and an agent in the companion computer. The IDL files are generated on the PX4 firmware build processed before being used on the agent code generators (with fastrtpsgen).
PX4-ROS2 bridge aka `px4_ros_com` = it's the same bridge as the previous one, the difference being the templates for the agent code are difference, because the types used for ROS2 are different from the types used for the "raw" RTPS/DDS applications. `px4_msgs` is used to generated those data types and the IDL files consumed by the agent code generator in `px4_ros_com` (also with fastrtpsgen).

Author: I am not entirely sure how to translate this to the documents (also where). Can you help?

# PX4-ROS 2 Bridge

This section of the guide focusses on the bridge between PX4 and ROS 2, made available through the *PX4-Fast RTPS(DDS) Bridge* and resulting products to interface with ROS 2, *px4_ros_com* and *px4_msgs* ROS 2 packages.

It also provides information in how to:
1. Connect ROS 2 nodes with PX4 (via the *microRTPS* Bridge, and using the `px4_ros_com` package)
1. Connect ROS (ROS "version 1") nodes with PX4 by additionally using the `ros1_bridge` package to bridge ROS 2 and ROS (1).

:::note
For information about using the *PX4-Fast RTPS(DDS) Bridge* without ROS 2, see the [RTPS/DDS Interface section](../middleware/micrortps.md).
:::

:::note
For a more detailed and visual explanation of using the bridge with ROS 2, see the following talks/presentation videos:
1. [PX4 Dev Summit 2019 - "ROS 2 Powered PX4"](https://www.youtube.com/watch?v=2Szw8Pk3Z0Q)
1. [ROS World 2020 - Getting started with ROS 2 and PX4](https://www.youtube.com/watch?v=qhLATrkA_Gw)
:::

## Why Two Bridges (PX4/ROS2 vs PX4/FastDDS)?

> The [PX4-Fast RTPS(DDS) Bridge](../middleware/micrortps.md) and PX4-ROS2 bridge (described here) are conceptually the same.

Reviewer: @TSC21 This is my attempt to capture the difference between the bridges, as discussed in https://github.com/PX4/PX4-user_guide/pull/1014/files#r569971259
The reason I think this is useful is because you otherwise don't have an overview of the application BUILD process - ie production of px4_msgs.

Please check this CAREFULLY. I am assuming that the client code is the same for both normal and ROS. The IDLs for ROS are generated to px4_msgs, which are built to ROS2 agent code using `px4_ros_com`.

Author: > Please check this CAREFULLY. I am assuming that the client code is the same for both normal and ROS.

Yes correct.

> The IDLs for ROS are generated to px4_msgs, which are built to ROS2 agent code using px4_ros_com.

No. px4_msgs generates the IDL codes used by the code generators in the px4_ros_com for the agent.

Reviewer: OK. let's be clear. This is what it says now:

> The PX4 build process automatically generates the IDL file-format required for ROS 2 and publishes them to `px4_msgs`.
> The `px4_msgs` are then processed by the agent code generator in `px4_ros_com` to create the agent source code.

You are saying that  `px4_msgs` generates IDL codes NOT PX4 build process. Further that these IDL codes are used by the code generators in the px4_ros_com to create the agent.

What/when does  `px4_msgs` use as the source for generating the IDL codes? I THINK from other discussions that these are essentially generated by the firmware as part of a build. Correct?

- They both have a client running on PX4 that communicates with an agent running on a companion computer.
- The PX4 firmware build process creates the client code.
- The PX4 firmware build process creates IDL files matching a select set of UORB topics that will be shared over DDS.
- These IDL files are then used by code generators to create the Agent source code, which can then be compiled for the companion. 

The difference is that ROS 2 uses slightly different DDS types than "raw" RTPS/DDS applications.
These types require slighty different agent code, generated from slightly different IDL files.
The client-side code is the same in both cases.

> The PX4 build process automatically generates the IDL file-format required for ROS 2 and publishes them to `px4_msgs`.

Author: ```suggestion
The PX4 build process creates ROS2 message files for each uORB message and deploys them to the `px4_msgs` repo. Based on those `.msg` files, the `px4_msgs` package build will generate the ROS 2 interface and typesupport code to be used by ROS2 nodes, but also simultaneously generates the IDL files that are going to be consumed by the code generators in `px4_ros_com` to generate the agent source code compatible with ROS 2.
```

Reviewer: I merged a slight variation of this. Thanks!

> The `px4_msgs` are then processed by the agent code generator in `px4_ros_com` to create the agent source code.

Author: ```suggestion
```

:::note 

> This means that you must use a version of `px4_msgs` that matches your firmware!

Reviewer: Also check this. It makes sense that you will have to have a set that matches what was built into firmware, but how do you match the version in PX4 v1.10 say?

Author: You don't, at least now. I need to automate the deployments for releases still.

> This means that you must use a version of `px4_msgs` that matches your firmware!

Author: ```suggestion
This means that you must use a version of `px4_msgs` that matches your firmware!

Note that by the time of writing there are no PX4 firmware release compatible `px4_ros_com` or `px4_msgs` (unless one looks for the commit hashes that match these repos with the PX4 firmware releases). It is in the plan of the development team to have this automated, in the sense that these repos will have branches which will be compatible with the respective PX4-Autopilot releases.
```
